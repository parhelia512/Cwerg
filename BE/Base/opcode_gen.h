#pragma once
// (c) Robert Muth - see LICENSE for more info
// NOTE: this file is PARTIALLY autogenerated via: ./opcode_tab.py gen_h

#include <cstdint>
#include <string_view>

namespace cwerg::base {

const unsigned MAX_OPERANDS = 5;
const unsigned MAX_PARAMETERS = 64;

/* @AUTOGEN-START@ */
enum class OPC : uint8_t {
    INVALID = 0x00,

    ADD = 0x10,
    SUB = 0x11,
    MUL = 0x12,
    DIV = 0x13,
    XOR = 0x18,
    AND = 0x19,
    OR = 0x1a,
    SHL = 0x1b,
    SHR = 0x1c,
    REM = 0x1d,
    CLMUL = 0x1e,

    BEQ = 0x20,
    BNE = 0x21,
    BLT = 0x22,
    BLE = 0x23,
    SWITCH = 0x28,
    BRA = 0x29,
    RET = 0x2a,
    BSR = 0x2b,
    JSR = 0x2c,
    SYSCALL = 0x2d,
    TRAP = 0x2e,

    PUSHARG = 0x30,
    POPARG = 0x31,
    CONV = 0x32,
    BITCAST = 0x33,
    MOV = 0x34,
    CMPEQ = 0x35,
    CMPLT = 0x36,
    LEA = 0x38,
    LEA_MEM = 0x39,
    LEA_STK = 0x3a,
    LEA_FUN = 0x3b,

    LD = 0x40,
    LD_MEM = 0x41,
    LD_STK = 0x42,
    ST = 0x44,
    ST_MEM = 0x45,
    ST_STK = 0x46,
    CAS = 0x48,
    CAS_MEM = 0x49,
    CAS_STK = 0x4a,

    CEIL = 0x50,
    FLOOR = 0x51,
    ROUND = 0x52,
    TRUNC = 0x53,
    COPYSIGN = 0x54,
    SQRT = 0x55,

    CNTLZ = 0x60,
    CNTTZ = 0x61,
    CNTPOP = 0x62,

    NOP = 0x70,
    NOP1 = 0x71,
    LINE = 0x77,
    INLINE = 0x78,
    GETFP = 0x79,
    GETSP = 0x7a,
    GETTP = 0x7b,

    DIR_MEM = 0x01,
    DIR_DATA = 0x02,
    DIR_ADDR_FUN = 0x03,
    DIR_ADDR_MEM = 0x04,
    DIR_FUN = 0x05,
    DIR_BBL = 0x06,
    DIR_REG = 0x07,
    DIR_STK = 0x08,
    DIR_JTB = 0x09,
};

enum class OPC_GENUS : uint8_t {
    INVALID = 0,
    BASE = 1,
    TBD = 2,
};

enum class FUN_KIND : uint8_t {
    INVALID = 0,
    BUILTIN = 1,
    EXTERN = 2,
    NORMAL = 3,
    SIGNATURE = 4,
};

enum class MEM_KIND : uint8_t {
    INVALID = 0,
    RO = 1,
    RW = 2,
    TLS = 3,
    FIX = 4,
    EXTERN = 5,
    BUILTIN = 6,
};

enum class TC : uint8_t {
    INVALID = 0,
    ANY = 1,
    ADDR_NUM = 2,
    ADDR_INT = 3,
    NUM = 4,
    FLT = 5,
    INT = 6,
    ADDR = 7,
    CODE = 8,
    UINT = 9,
    SINT = 10,
    OFFSET = 11,
    SAME_AS_PREV = 20,
    SAME_SIZE_AS_PREV = 22,
};

enum class OPC_KIND : uint8_t {
    INVALID = 0,
    ALU = 1,
    ALU1 = 2,
    MOV = 3,
    LEA = 4,
    LEA1 = 5,
    COND_BRA = 6,
    BRA = 7,
    BSR = 8,
    JSR = 9,
    SWITCH = 10,
    RET = 11,
    SYSCALL = 12,
    ST = 13,
    LD = 14,
    PUSHARG = 15,
    POPARG = 16,
    NOP = 17,
    NOP1 = 18,
    CONV = 19,
    CMP = 20,
    BCOPY = 21,
    BZERO = 22,
    CAS = 23,
    GETSPECIAL = 24,
    INLINE = 25,
    LINE = 26,
    DIRECTIVE = 27,
};

enum class DK : uint8_t {
    INVALID = 0,
    S8 = 16,
    S16 = 17,
    S32 = 18,
    S64 = 19,
    U8 = 32,
    U16 = 33,
    U32 = 34,
    U64 = 35,
    F8 = 48,
    F16 = 49,
    F32 = 50,
    F64 = 51,
    A32 = 66,
    A64 = 67,
    C32 = 82,
    C64 = 83,
};

enum class OP_KIND : uint8_t {
    INVALID = 0,
    REG = 1,
    CONST = 2,
    REG_OR_CONST = 3,
    BBL = 4,
    MEM = 5,
    STK = 6,
    FUN = 7,
    JTB = 8,
    TYPE_LIST = 20,
    DATA_KIND = 21,
    MEM_KIND = 23,
    FUN_KIND = 24,
    FIELD = 25,
    NAME = 26,
    NAME_LIST = 27,
    INT = 28,
    BBL_TAB = 29,
    BYTES = 30,
};

enum OA : uint16_t {
    BBL_TERMINATOR = 1,
    NO_FALL_THROUGH = 2,
    CALL = 4,
    COMMUTATIVE = 8,
    MEM_RD = 16,
    MEM_WR = 32,
    SPECIAL = 64,
};
/* @AUTOGEN-END@ */

constexpr const uint8_t DK_FLAVOR_S = 1 << 4;
constexpr const uint8_t DK_FLAVOR_U = 2 << 4;
constexpr const uint8_t DK_FLAVOR_F = 3 << 4;
constexpr const uint8_t DK_FLAVOR_A = 4 << 4;
constexpr const uint8_t DK_FLAVOR_C = 5 << 4;

inline int DKFlavor(DK rk) { return uint8_t(rk) & 0x70; }
inline int DKBitWidth(DK rk) { return 8 << (uint8_t(rk) & 0x7); }

struct Opcode {
  // layout optimized for usage
  const OP_KIND operand_kinds[MAX_OPERANDS];
  const OPC_KIND kind;
  const OPC_GENUS genus;
  const uint8_t num_operands;
  const uint8_t num_defs;  // defs are always the first few operands
  const TC constraints[MAX_OPERANDS];
  const char* const name;
  uint16_t attributes;

  bool HasAttribute(OA flags) const { return (attributes & flags) != 0; }

  bool IsCall() const { return HasAttribute(OA::CALL); }

  bool HasSideEffect() const {
    return HasAttribute(OA(OA::BBL_TERMINATOR | OA::CALL | OA::MEM_RD |
                           OA::MEM_WR | OA::SPECIAL));
  }

  bool IsBblTerminator() const { return HasAttribute(OA::BBL_TERMINATOR); }
  bool HasFallthrough() const { return !HasAttribute(OA::NO_FALL_THROUGH); }
};

extern const Opcode GlobalOpcodes[256];

extern OPC OPCFromString(std::string_view name);
extern OPC_GENUS OpcGenusFromString(std::string_view name);
extern FUN_KIND FKFromString(std::string_view name);
extern MEM_KIND MKFromString(std::string_view name);
extern TC TCFromString(std::string_view name);
extern DK DKFromString(std::string_view name);

extern const char* EnumToString(OPC_GENUS x);
extern const char* EnumToString(FUN_KIND x);
extern const char* EnumToString(MEM_KIND x);
extern const char* EnumToString(TC x);
extern const char* EnumToString(DK x);
extern const char* EnumToString(OP_KIND x);
inline const char* EnumToString(OPC x) {
  return GlobalOpcodes[unsigned(x)].name;
}

}  // namespace cwerg::base
