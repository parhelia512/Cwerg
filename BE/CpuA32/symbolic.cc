// (c) Robert Muth - see LICENSE for more info

#include "BE/CpuA32/symbolic.h"
#include "Util/assert.h"
#include "Util/parse.h"

#include <cstring>
#include <string_view>

namespace cwerg::a32 {
namespace {

char* strappenddec(char* dst, uint32_t n) {
  if (n <= 0x7fffffff) {
    return dst + ToDecString(n, dst).size();
  } else {
    *dst++ = '-';
    return dst + ToDecString(~n + 1, dst).size();
  }
}

char* strappendhex(char* dst, uint32_t n) {
  return dst + ToHexString(n, dst).size();
}

char* strappend(char* dst, std::string_view src) {
  std::memcpy(dst, src.data(), src.size());
  dst[src.size()] = 0;
  return dst + src.size();
}

char* SymbolizeOperand(char* buffer, uint32_t data, OK ok) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  const uint32_t x = a32::DecodeOperand(data, ok);
  switch (fi.kind) {
    default:
    case FK::NONE:
      ASSERT(false, "unreachable");
      return nullptr;
    case FK::INT_SIGNED_CUSTOM:
      return strappenddec(buffer, x);
    case FK::FLT_CUSTOM:
      ASSERT(x == 0, "");
      return strappend(buffer, "0.0");
    case FK::LIST:
      ASSERT(x < fi.num_names, "out of range for list " << x << " " << (int)ok);
      return strappend(buffer, fi.names[x]);
    case FK::INT_SIGNED:
      return strappenddec(buffer, x);
    case FK::INT:
      buffer = strappend(buffer, fi.prefix);
      return strappenddec(buffer, x);
    case FK::INT_HEX:
      buffer = strappend(buffer, fi.prefix);
      buffer = strappend(buffer, "0x");
      return strappendhex(buffer, x);
  }
}

bool has_prefix(std::string_view name, std::string_view prefix) {
  return name.substr(0, prefix.size()) == prefix;
}

uint32_t UnsymbolizeOperandString(const FieldInfo& fi, std::string_view s) {
  switch (fi.kind) {
    default:
    case FK::NONE:
      ASSERT(false, "unreachable");
      return kEncodeFailure;
    case FK::INT_SIGNED:
    case FK::INT_SIGNED_CUSTOM: {
      auto val = ParseInt<int32_t>(s);
      if (!val) return kEncodeFailure;
      return val.value();
    }
    case FK::FLT_CUSTOM:
      if (s != "0.0" && s != ".0" && s != "0") return kEncodeFailure;
      return 0;  // this is really the bit representation of float(0.0)
    case FK::LIST:
      for (unsigned i = 0; i < fi.num_names; ++i) {
        if (s == fi.names[i]) return i;
      }
      return kEncodeFailure;
    case FK::INT:
    case FK::INT_HEX: {
      auto val = ParseInt<uint32_t>(s);
      if (!val) return kEncodeFailure;
      return val.value();
    }
  }
}

uint32_t UnsymbolizeOperand(OK ok, std::string_view s) {
  const FieldInfo& fi = FieldInfoTable[uint8_t(ok)];
  if (!fi.prefix.empty() && has_prefix(s, fi.prefix)) {
    s.remove_prefix(fi.prefix.size());
  } else if (s == "r10")
    s = "sl";
  else if (s == "r11")
    s = "fp";
  else if (s == "r12")
    s = "ip";

  return a32::EncodeOperand(UnsymbolizeOperandString(fi, s), ok);
}

const constexpr struct {
  std::string_view expr_kind;
  elf::RELOC_TYPE_ARM reloc_kind;
  bool local;
} kExprToReloc[] = {
    // order by frequency
    {"jump24", elf::RELOC_TYPE_ARM::JUMP24, true},
    {"call", elf::RELOC_TYPE_ARM::CALL, false},
    {"abs32", elf::RELOC_TYPE_ARM::ABS32, false},
    {"movw_abs_nc", elf::RELOC_TYPE_ARM::MOVW_ABS_NC, false},
    {"movt_abs", elf::RELOC_TYPE_ARM::MOVT_ABS, false},
    {"loc_movw_abs_nc", elf::RELOC_TYPE_ARM::MOVW_ABS_NC, true},
    {"loc_movt_abs", elf::RELOC_TYPE_ARM::MOVT_ABS, true},
};

bool HandleRelocation(std::string_view expr, unsigned pos, Ins* ins) {
  auto expr_op = ParseExpressionOp(expr);
  if (!expr_op) return false;
  ins->operands[pos] = expr_op->offset;
  for (const auto& x : kExprToReloc) {
    if (x.expr_kind == expr_op->reloc_name) {
      ins->set_reloc(x.reloc_kind, x.local, pos, expr_op->symbol_name);
      return true;
    }
  }
  return false;
}

void SymbolizeReloc(char* cp, const Ins& ins, uint32_t addend) {
  cp = strappend(cp, "expr:");
  switch (ins.reloc_kind) {
    case elf::RELOC_TYPE_ARM::JUMP24:
      ASSERT(ins.is_local_sym, "");
      cp = strappend(cp, "jump24:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::CALL:
      cp = strappend(cp, "call:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::MOVT_ABS:
      cp = strappend(cp, ins.is_local_sym ? "loc_movt_abs:" : "movt_abs:");
      cp = strappend(cp, ins.reloc_symbol);
      break;
    case elf::RELOC_TYPE_ARM::MOVW_ABS_NC:
      cp =
          strappend(cp, ins.is_local_sym ? "loc_movw_abs_nc:" : "movw_abs_nc:");
      cp = strappend(cp, ins.reloc_symbol);

      break;
    default:
      ASSERT(false, "");
  }
  if (addend != 0) {
    *cp++ = ':';
    cp = strappenddec(cp, addend);
  }
}

}  // namespace

std::string_view InsSymbolize(const Ins& ins, std::vector<std::string>* ops) {
  char buffer[128];
  for (unsigned i = 0; i < ins.opcode->num_fields; ++i) {
    if (ins.has_reloc() && i == ins.reloc_pos) {
      SymbolizeReloc(buffer, ins, ins.operands[i]);
    } else {
      SymbolizeOperand(buffer, ins.operands[i], ins.opcode->fields[i]);
    }
    ops->emplace_back(buffer);
  }
  return ins.opcode->enum_name;
}

bool InsFromSymbolized(const std::vector<std::string_view>& token, Ins* ins) {
  ins->opcode = FindOpcodeForMnemonic(token[0]);
  if (ins->opcode == nullptr) {
    std::cerr << "unknown opcode " << token[0] << "\n";
    return false;
  }
  uint32_t operand_count = 0;
  // CodeGenA32 relies on this
  if (token.size() == ins->opcode->num_fields) {
    ins->operands[operand_count++] = 14;  // predicate 'al'
  }
  ASSERT(token.size() - 1 + operand_count == ins->opcode->num_fields, "");
  for (unsigned i = 1; i < token.size(); ++i, ++operand_count) {
    if (token[i].substr(0, 5) == "expr:") {
      if (!HandleRelocation(token[i].substr(5), operand_count, ins)) {
        std::cerr << "malformed relocation expression " << token[i] << "\n";
        return false;
      }
    } else {
      const OK ok = ins->opcode->fields[operand_count];
      auto val = UnsymbolizeOperand(ok, token[i]);
      if (val == kEncodeFailure) {
        std::cerr << "cannot parse " << token[i] << " "
                  << "\n";
        return false;
      }
      ins->operands[operand_count] = val;
    }
  }
  return true;
}

std::string SymbolizeRegListMask(uint32_t mask) {
  std::string regs = "{";
  std::string_view sep;
  for (unsigned int i = 0; i < 16; ++i) {
    if (mask & (1U << i)) {
      regs += sep;
      regs += EnumToString(a32::REG(i));
      sep = ", ";
    }
  }
  regs += "}";
  return regs;
}

std::string SymbolizeRegRange(uint32_t start_reg_no,
                              uint32_t count,
                              bool single) {
  ASSERT(count > 0, "");
  std::string_view start_reg = single ? EnumToString(a32::SREG(start_reg_no))
                                      : EnumToString(a32::DREG(start_reg_no));
  std::string out = "{";
  out += start_reg;
  if (count > 1) {
    uint32_t end_reg_no = start_reg_no + count - 1;
    std::string_view end_reg = single ? EnumToString(a32::SREG(end_reg_no))
                                      : EnumToString(a32::DREG(end_reg_no));
    out += "-";
    out += end_reg;
  }
  return out + "}";
}

}  // namespace cwerg::a32
