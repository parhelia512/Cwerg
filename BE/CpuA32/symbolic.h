#pragma once
// (c) Robert Muth - see LICENSE for more info

#include <optional>
#include <string_view>
#include <vector>
#include "BE/CpuA32/opcode_gen.h"

namespace cwerg::a32 {

extern std::string_view InsSymbolize(const Ins& ins, std::vector<std::string>* ops);

extern bool InsFromSymbolized(const std::vector<std::string_view>& token, Ins* ins);

extern std::string SymbolizeRegListMask(uint32_t mask);

extern std::string SymbolizeRegRange(uint32_t start_reg_no,
                                     uint32_t count,
                                     bool single);
}  // namespace cwerg::a32
