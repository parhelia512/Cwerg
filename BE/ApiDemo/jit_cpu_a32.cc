// This examples shows how to jit a32 assembly directly (not cwerg IR)
#include "BE/CpuA32/opcode_gen.h"
#include "BE/CpuA32/symbolic.h"
#include "Util/assert.h"

#include <sys/mman.h>
#include <iostream>

using namespace cwerg::a32;

// +-prefix converts an enum the underlying type
template <typename T>
constexpr auto operator+(T e) noexcept
    -> std::enable_if_t<std::is_enum<T>::value, std::underlying_type_t<T>> {
  return static_cast<std::underlying_type_t<T>>(e);
}

using FunPtr = uint32_t (*)(uint32_t);

void DumpIns(uint32_t data) {
  Ins ins;
  Disassemble(&ins, data);
  std::vector<std::string> ops;
  std::string_view enum_name = InsSymbolize(ins, &ops);
  std::cerr << "0x" << std::hex << data << " " << enum_name;
  std::string_view sep = " ";
  for (const std::string& op : ops) {
    std::cerr << sep << op;
    sep = ", ";
  }
  std::cerr << "\n";
}

// encode a OK.SIMM_0_23
constexpr uint32_t encode_branch_offset(int32_t offset) {
  return offset & 0xffffff;
}

#define OPCODE(o) &OpcodeTable[+OPC::o]

int main(int argc, char* argv[]) {
  const Ins Fibonacci[] = {
      //
      {OPCODE(stmdb_update), {+PRED::al, +REG::sp, 0x4030}},
      {OPCODE(cmp_imm), {+PRED::al, +REG::r0, 1}},
      {OPCODE(b), {+PRED::le, encode_branch_offset(7)}},
      //
      {OPCODE(mov_imm), {+PRED::al, +REG::r4, 0}},
      {OPCODE(mov_regimm), {+PRED::al, +REG::r5, +REG::r0, +SHIFT::lsl, 0}},
      //
      {OPCODE(sub_imm), {+PRED::al, +REG::r0, +REG::r5, 1}},
      {OPCODE(bl), {+PRED::al, encode_branch_offset(-8)}},
      {OPCODE(add_regimm),
       {+PRED::al, +REG::r4, +REG::r4, +REG::r0, +SHIFT::lsl, 0}},
      //
      {OPCODE(sub_imm), {+PRED::al, +REG::r0, +REG::r5, 2}},
      {OPCODE(bl), {+PRED::al,encode_branch_offset(-11)}},
      {OPCODE(add_regimm),
       {+PRED::al, +REG::r0, +REG::r4, +REG::r0, +SHIFT::lsl, 0}},
      //
      {OPCODE(ldmia_update), {+PRED::al, 0x8030, +REG::sp}}

  };

  uint32_t* memory =
      (uint32_t*)mmap(nullptr, 4096, PROT_READ | PROT_WRITE | PROT_EXEC,
                      MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
  ASSERT(memory != nullptr, "");
  unsigned i = 0;
  for (const auto& ins : Fibonacci) {
    memory[i] = Assemble(ins);
    DumpIns(memory[i]);
    ++i;
  }

  FunPtr f = reinterpret_cast<FunPtr>(memory);
  for (unsigned i = 0; i < 10; ++i) {
    std::cout << std::dec << i << " " << f(i) << "\n";
  }
  return 0;
}
