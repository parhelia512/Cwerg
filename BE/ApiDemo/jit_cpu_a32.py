#!/bin/env python3


import ctypes
import mmap
import platform
import struct
import sys

from BE.CpuA32 import symbolic as dis
from BE.CpuA32 import opcode_tab as a32


class TestBuffer:

    def __init__(self):
        self.instructions = []

    def write(self, bs):
        self.instructions.append(bs)

    def dump(self):
        for i, bs in enumerate(self.instructions):
            print(f"{i:2d} {bs}")

    def dump_fancy(self):
        for i, bs in enumerate(self.instructions):
            data = struct.unpack("<I", bs)[0]
            ins = a32.Disassemble(data)
            _, ops = dis.InsSymbolize(ins)
            if ops and ops[0] == "al":
                ops.pop(0)
            print(
                f"{i:2d} {[f'{b:02x}' for b in bs]} {data:08x} {ins.opcode.name} {' '.join(ops)}")


def MakeBuffer(size, prot):
    buf = mmap.mmap(-1, size, prot=prot)
    void_pointer = ctypes.c_void_p.from_buffer(buf)
    return buf, ctypes.addressof(void_pointer)


def EmitNull(code_buf):
    for ins in [
        # "al" predicate is optional
        dis.InsFromSymbolized("add_imm", ["r0", "r0", "1"]),
        dis.InsFromSymbolized("mov_regimm", ["pc", "lr", "lsl", "0"])
    ]:
        code_buf.write(a32.Assemble(ins).to_bytes(4, "little"))


def EmitMul(code_buf):
    for ins in [
        # "al" predicate is optional
        dis.InsFromSymbolized("mul", ["r0", "r1", "r0"]),
        dis.InsFromSymbolized("mov_regimm", ["pc", "lr", "lsl", "0"])
    ]:
        code_buf.write(a32.Assemble(ins).to_bytes(4, "little"))


def EmitFib(code_buf):
    for ins in [
        # e92d4030 stm sp!, {r4,r5,lr}
        dis.InsFromSymbolized("stmdb_update", ["al", "sp", "reglist:16432"]),
        dis.InsFromSymbolized("cmp_imm", ["al", "r0", "1"]),
        dis.InsFromSymbolized("b", ["le", "7"]),
        dis.InsFromSymbolized("mov_imm", ["al", "r4", "0"]),
        dis.InsFromSymbolized("mov_regimm", ["al", "r5", "r0", "lsl", "0"]),
        #
        dis.InsFromSymbolized("sub_imm", ["al", "r0", "r5", "1"]),
        dis.InsFromSymbolized("bl", ["al", "-8"]),
        dis.InsFromSymbolized(
            "add_regimm", ["al", "r4", "r4", "r0", "lsl", "0"]),
        # #
        dis.InsFromSymbolized("sub_imm", ["al", "r0", "r5", "2"]),
        dis.InsFromSymbolized("bl", ["al", "-11"]),
        dis.InsFromSymbolized(
            "add_regimm", ["al", "r0", "r4", "r0", "lsl", "0"]),
        # e8bd4030 ldm sp!, {r4,r5,pc}
        dis.InsFromSymbolized("ldmia_update", ["al", "reglist:32816", "sp"]),
    ]:
        code_buf.write(a32.Assemble(ins).to_bytes(4, "little"))


def gen_code(x):
    isa = platform.machine()
    if isa != "armv7l":
        print(f"\nIncompatible machine architecture {isa}: no execution")
        return

    code_buf, code_addr = MakeBuffer(
        mmap.PAGESIZE,
        mmap.PROT_READ | mmap.PROT_WRITE | mmap.PROT_EXEC)

    data_buf, data_addr = MakeBuffer(
        mmap.PAGESIZE,
        mmap.PROT_READ | mmap.PROT_WRITE)

    # print(f"CODE-BUFFER: {code_buf} {code_addr:08x}")
    ftype = ctypes.CFUNCTYPE(ctypes.c_int, ctypes.c_int)
    f = ftype(code_addr)

    EmitFib(code_buf)
    print("Running Code")
    r = f(x)
    print(r)

    # del fpointer
    code_buf.close()


if len(sys.argv) > 1:
    gen_code(int(sys.argv[1]))
else:
    print("Code generator dry run - no execution")
    for snippet in [EmitNull, EmitMul, EmitFib]:
        print(f"\n{snippet.__name__}")
        test_buffer = TestBuffer()
        snippet(test_buffer)
        test_buffer.dump_fancy()
