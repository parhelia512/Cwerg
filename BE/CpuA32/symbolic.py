"""
Convert ARM32 instruction into human readable form suitable to
be processed by an assembler.
"""
import struct
from typing import List, Tuple, Set

from BE.CpuA32 import opcode_tab as a32
from BE.Elf import enum_tab


def _EmitReloc(ins: a32.Ins, pos: int) -> str:
    if ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.JUMP24:
        assert ins.is_local_sym, f"expected local symbol"
        return f"expr:jump24:{ins.reloc_symbol}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVT_ABS:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movt_abs:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC:
        loc = "loc_" if ins.is_local_sym else ""
        offset = "" if ins.operands[pos] == 0 else f":{ins.operands[pos]}"
        return f"expr:{loc}movw_abs_nc:{ins.reloc_symbol}{offset}"
    elif ins.reloc_kind == enum_tab.RELOC_TYPE_ARM.CALL:
        return f"expr:call:{ins.reloc_symbol}"
    else:
        assert False


def _FloatFrom32BitRepresentation(data: int) -> float:
    return struct.unpack('<f', int.to_bytes(data, 4, "little"))[0]


def _SymbolizeOperand(ok: a32.OK, data: int) -> str:
    """Convert an operand in integer form as found in  arm.Ins to a string

    (This does not handle relocation expressions.)
    """
    data = a32.DecodeOperand(ok, data)
    assert data >= 0, f"expected unsigned for {ok}: {data}"
    t = a32.FIELD_DETAILS[ok]
    if t.kind == a32.FK.LIST:
        return t.names[data]
    elif t.kind == a32.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _FloatFrom32BitRepresentation(data)
        return str(data)
    elif t.kind == a32.FK.INT_SIGNED_CUSTOM or t.kind == a32.FK.INT_SIGNED:
        # we only care about the signed aspect
        if data >= 1 << 31:
            data -= (1 << 32)
        return str(data)
    elif t.kind == a32.FK.INT_HEX:
        return t.prefix + hex(data)
    else:
        return t.prefix + str(data)


_REG_REWRITES = {
    "r10": "sl",
    "r11": "fp",
    "r12": "ip",
    "r14": "lr",
}


def _FloatTo32BitRepresentation(num: float) -> int:
    b = struct.pack('<f', num)
    assert len(b) == 4
    return int.from_bytes(b, "little")


def _UnsymbolizeOperand(ok: a32.OK, op: str) -> int:
    """Convert a symbolized operand  into an int suitable for arm.Ins

    This does not handle relocation expressions.
    """
    t = a32.FIELD_DETAILS[ok]
    if t.prefix and op.startswith(t.prefix):
        op = op[len(t.prefix):]
    if t.kind == a32.FK.LIST:
        op = _REG_REWRITES.get(op, op)
        data = t.names.index(op)
    elif t.kind == a32.FK.FLT_CUSTOM:
        # we only care about the float aspect
        data = _FloatTo32BitRepresentation(float(op))
    elif t.kind == a32.FK.INT_SIGNED_CUSTOM or t.kind == a32.FK.INT_SIGNED:
        # we only care about the signed aspect
        data = int(op, 0)
        if data < 0:
            data += 1 << 32
    else:
        data = int(op, 0)
    assert data is not None and data >= 0, f"bad data {ok}: {op}"
    return a32.EncodeOperand(ok, data)


_RELOC_KIND_MAP = {
    "jump24": (enum_tab.RELOC_TYPE_ARM.JUMP24, True),  # always local
    #
    "abs32": (enum_tab.RELOC_TYPE_ARM.ABS32, False),
    "call": (enum_tab.RELOC_TYPE_ARM.CALL, False),
    "movw_abs_nc": (enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC, False),
    "movt_abs": (enum_tab.RELOC_TYPE_ARM.MOVT_ABS, False),
    #
    "loc_abs32": (enum_tab.RELOC_TYPE_ARM.ABS32, True),
    "loc_call": (enum_tab.RELOC_TYPE_ARM.CALL, True),
    "loc_movw_abs_nc": (enum_tab.RELOC_TYPE_ARM.MOVW_ABS_NC, True),
    "loc_movt_abs": (enum_tab.RELOC_TYPE_ARM.MOVT_ABS, True),
}

_RELOC_OK: Set[a32.OK] = {a32.OK.SIMM_0_23, a32.OK.IMM_0_11_16_19}


def InsSymbolize(ins: a32.Ins) -> Tuple[str, List[str]]:
    """Convert all the operands in an arm.Ins to strings including relocs
    """
    ops = []
    for pos, (ok, value) in enumerate(zip(ins.opcode.fields, ins.operands)):
        if ok in _RELOC_OK and ins.has_reloc() and ins.reloc_pos == pos:
            # Note: we essentially store the "addend" here
            ops.append(_EmitReloc(ins, pos))
        else:
            ops.append(_SymbolizeOperand(ok, value))

    return ins.opcode.name, ops


def InsFromSymbolized(mnemonic, token: List[str]) -> a32.Ins:
    """Takes textual form of an A32 instruction (in systematic notation) and parses it into a Ins

     * Supports relocatable expressions
     * Adds missing "al" predicate if necessary
     Example input:
     "add_regimm", ["r4", "r4", "r0", "lsl", "0"]
     """
    opcode: a32.Opcode = a32.Opcode.name_to_opcode[mnemonic]
    # CodeGenA32 relies on this
    if opcode.HasPred() and len(token) == len(opcode.fields) - 1:
        token = ["al"] + token
    ins = a32.Ins(opcode)
    assert len(token) == len(opcode.fields), f"expected {len(opcode.fields)} operands in {mnemonic} {token}"
    for pos, (t, ok) in enumerate(zip(token, opcode.fields)):
        if t.startswith("expr:"):
            # expr strings have the form expr:<rel-kind>:<symbol>:<addend>, e.g.:
            #   expr:movw_abs_nc:string_pointers:5
            #   expr:call:putchar
            rel_token = t.split(":")

            ins.set_reloc(*_RELOC_KIND_MAP[rel_token[1]], pos, rel_token[2])
            addend = 0 if len(rel_token) == 3 else int(rel_token[3], 0)
            ins.operands.append(addend)
        else:
            ins.operands.append(_UnsymbolizeOperand(ok, t))
    return ins
