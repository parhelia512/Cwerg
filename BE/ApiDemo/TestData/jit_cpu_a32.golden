Code generator dry run - no execution

EmitNull
 0 ['01', '00', '80', 'e2'] e2800001 add_imm r0 r0 1
 1 ['0e', 'f0', 'a0', 'e1'] e1a0f00e mov_regimm pc lr lsl 0

EmitMul
 0 ['91', '00', '00', 'e0'] e0000091 mul r0 r1 r0
 1 ['0e', 'f0', 'a0', 'e1'] e1a0f00e mov_regimm pc lr lsl 0

EmitFib
 0 ['30', '40', '2d', 'e9'] e92d4030 stmdb_update sp reglist:0x4030
 1 ['01', '00', '50', 'e3'] e3500001 cmp_imm r0 1
 2 ['07', '00', '00', 'da'] da000007 b le 7
 3 ['00', '40', 'a0', 'e3'] e3a04000 mov_imm r4 0
 4 ['00', '50', 'a0', 'e1'] e1a05000 mov_regimm r5 r0 lsl 0
 5 ['01', '00', '45', 'e2'] e2450001 sub_imm r0 r5 1
 6 ['f8', 'ff', 'ff', 'eb'] ebfffff8 bl -8
 7 ['00', '40', '84', 'e0'] e0844000 add_regimm r4 r4 r0 lsl 0
 8 ['02', '00', '45', 'e2'] e2450002 sub_imm r0 r5 2
 9 ['f5', 'ff', 'ff', 'eb'] ebfffff5 bl -11
10 ['00', '00', '84', 'e0'] e0840000 add_regimm r0 r4 r0 lsl 0
11 ['30', '80', 'bd', 'e8'] e8bd8030 ldmia_update reglist:0x8030 sp
